src/MSBuild/TerminalLogger/TerminalLogger.cs

```diff
@@ -701,11 +701,8 @@ private void TargetStarted(object sender, TargetStartedEventArgs e)
 
     private void UpdateNodeStatus(BuildEventContext buildEventContext, NodeStatus? nodeStatus)
     {
-        lock (_lock)
-        {
-            int nodeIndex = NodeIndexForContext(buildEventContext);
-            _nodes[nodeIndex] = nodeStatus;
-        }
+        int nodeIndex = NodeIndexForContext(buildEventContext);
+        _nodes[nodeIndex] = nodeStatus;
```

Comment:
nit: Although from memory safety point of view [this is safe](https://github.com/dotnet/runtime/blob/main/docs/design/specs/Memory-model.md#object-assignment), I feel like it does make the code more susceptible to state tearing and other concurrency bugs. Would you consider using `Volatile.Write` to _document_ the fact that this is writing to a location which is read by other threads concurrently?

Reply:
I don't think `Volatile.Write` is the right mechanism here as it seems to impact ordering of other operations around this, not this operation itself.

I added a comment to the `_nodes` field.

Reply:
I think that's intended. To me the gist of the problem in patterns like this is whether all writes that happened _before_ this one were actually flushed/committed/published - whatever the right term is - to other processors. And if .NET implemented only the memory model as specified in ECMA, on hardware like ARM64 without a memory barrier this could actually misbehave despite the write being atomic. Another processor could read the new reference from the `_nodes` array but the memory pointed to by the reference could still have old contents in it because the writes performed as part of constructing the `NodeStatus` were not properly propagated.